src/Storages/MergeTree/MutateTask.cpp

```diff
@@ -1470,7 +1470,7 @@ bool MutateTask::execute()
             if (task->executeStep())
                 return true;
 
-            promise.set_value(ctx->new_data_part);
+            promise.set_value(std::move(ctx->new_data_part));
```

Comment:
I don't understand... if we don't move this data part and an exception is thrown, who and why someone will continue to hold it?

Note: I did not read all the code, only the diff.